Implement creating host with encryption connect, update host status and delete hosts

judu opened the pull request:
# Changes

## Add TlsConfig to host.create

This allows someone to create a host with tls_connect/accept.
It supports psk and certificate.

This implementation is opinionated. I prefer to force the user of the lib to choose one of None, Some(Psk) or Some(Cert) and to setup tls_accept to the same as tls_connect.

My opinion is it’s better to guide the user with a simple API rather than having to manually set:

```
tls_connect: 1 | 2 | 4,
tls_accept: 1..7
tls_subject: …
tls_issuer: …
tls_psk_identity: …
tls_psk: …
```

*I’m open to change that part!*

## Implement host.update and host.delete

On the host.update part: I only implemented the status param as I don’t need anything else. I’m open to implement a bit more, but most of it will be unused :innocent:

judu committed:
fix(tests): Ensure the faker creates long enough passwords

judu committed:
feat(host.update): implement update of host status

I don’t need to update anything else, so I figured this should be enough
for now.

judu committed:
feat(host.create): implement TlsConfig

This allows someone to create a host with tls_connect/accept.
It supports psk and certificate.

judu committed:
Bump deps

judu committed:
feat(host.delete): implement host delete operation

judu committed:
fix(test): feature v6 needs v7 to be disabled

lebe-dev commented:
Thanks for contribution, let me check

lebe-dev commented on src/tests/builder.rs:
```diff
@@ -88,7 +89,7 @@ impl TestEnvBuilder {
     }
 
-    pub fn create_host(&mut self, name: &str) -> &mut Self {
+    pub fn create_host(&mut self, name: &str, encryption_config: Option<TlsConfig>) -> &mut Self {
```
@judu imo, argument should be named as `tls_config`

lebe-dev commented:
I've found an issue for test against API v7. After you spin up fresh zabbix 7.x.x instance:

```
(base) ➜  pr-17 git:(push-lktkluuruklo) ./run-integration-tests-v7.sh
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.26s
     Running unittests src/lib.rs (target/debug/deps/zabbix_api-0c7e948664e6c4b5)

running 21 tests
test client::client::tests::get_api_info ... ok
test client::client::tests::create_host_with_cert ... FAILED
test client::client::tests::create_host_group_and_host ... ok
test client::client::tests::create_host_with_psk ... ok
test client::client::tests::create_item ... ok
test client::client::tests::create_trigger ... ok
test client::client::tests::create_user_group ... ok
test client::client::tests::create_web_scenario ... ok
test client::client::tests::get_host_groups_test ... ok
test client::client::tests::create_and_update_host ... ok
test client::client::tests::session_should_be_returned ... ok
test trigger::model::trigger_tests::deserialize_test ... ok
test client::client::tests::raw_api_call_test ... ok
test client::client::tests::get_hosts_test ... ok
test client::client::tests::get_triggers_test ... ok
test client::client::tests::get_items_test ... ok

(base) ➜  pr-17 git:(push-lktkluuruklo) ./run-integration-tests-v7.sh
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.26s
     Running unittests src/lib.rs (target/debug/deps/zabbix_api-0c7e948664e6c4b5)

running 21 tests
test client::client::tests::get_api_info ... ok
test client::client::tests::create_host_with_cert ... FAILED
test client::client::tests::create_host_group_and_host ... ok
test client::client::tests::create_host_with_psk ... ok
test client::client::tests::create_item ... ok
test client::client::tests::create_trigger ... ok
test client::client::tests::create_user_group ... ok
test client::client::tests::create_web_scenario ... ok
test client::client::tests::get_host_groups_test ... ok
test client::client::tests::create_and_update_host ... ok
test client::client::tests::session_should_be_returned ... ok
test trigger::model::trigger_tests::deserialize_test ... ok
test client::client::tests::raw_api_call_test ... ok
test client::client::tests::get_hosts_test ... ok
test client::client::tests::get_triggers_test ... ok
test client::client::tests::get_items_test ... ok
test client::client::tests::get_webscenarios_test ... ok
test client::client::user_tests::get_user_groups_test ... ok
test client::client::user_tests::get_users_test ... ok
test client::client::tests::delete_hosts_test ... ok
test client::client::user_tests::create_user_test ... ok

failures:

---- client::client::tests::create_host_with_cert stdout ----
[2025-05-28T15:22:34Z INFO  zabbix_api::client::client] getting auth session for user 'Admin'..
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] send post request to 'http://localhost:3080/api_jsonrpc.php'
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] ---[HTTP REQUEST]----
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] {"jsonrpc":"2.0","method":"user.login","params":{"password":"zabbix","username":"Admin"},"id":1}
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] ---[/HTTP REQUEST]----
[2025-05-28T15:22:34Z DEBUG reqwest::connect] starting new connection: http://localhost:3080/
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] ---[HTTP RESPONSE]----
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] {"jsonrpc":"2.0","result":"0930a6ef7dfe5d557e2ba863cabd2018","id":1}
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] ---[/HTTP RESPONSE]----
[2025-05-28T15:22:34Z INFO  zabbix_api::client::client] auth ok
[2025-05-28T15:22:34Z INFO  zabbix_api::client::client] creating host group 'oLJ0NxqbJTzoCAcgMIvYYo92QpSHyK'..
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] send post request to 'http://localhost:3080/api_jsonrpc.php'
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] ---[HTTP REQUEST]----
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] {"jsonrpc":"2.0","method":"hostgroup.create","params":{"name":"oLJ0NxqbJTzoCAcgMIvYYo92QpSHyK"},"id":1}
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] ---[/HTTP REQUEST]----

(base) ➜  pr-17 git:(push-lktkluuruklo) ./run-integration-tests-v7.sh
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.26s
     Running unittests src/lib.rs (target/debug/deps/zabbix_api-0c7e948664e6c4b5)

running 21 tests
test client::client::tests::get_api_info ... ok
test client::client::tests::create_host_with_cert ... FAILED
test client::client::tests::create_host_group_and_host ... ok
test client::client::tests::create_host_with_psk ... ok
test client::client::tests::create_item ... ok
test client::client::tests::create_trigger ... ok
test client::client::tests::create_user_group ... ok
test client::client::tests::create_web_scenario ... ok
test client::client::tests::get_host_groups_test ... ok
test client::client::tests::create_and_update_host ... ok
test client::client::tests::session_should_be_returned ... ok
test trigger::model::trigger_tests::deserialize_test ... ok
test client::client::tests::raw_api_call_test ... ok
test client::client::tests::get_hosts_test ... ok
test client::client::tests::get_triggers_test ... ok
test client::client::tests::get_items_test ... ok
test client::client::tests::get_webscenarios_test ... ok
test client::client::user_tests::get_user_groups_test ... ok
test client::client::user_tests::get_users_test ... ok
test client::client::tests::delete_hosts_test ... ok
test client::client::user_tests::create_user_test ... ok

failures:

---- client::client::tests::create_host_with_cert stdout ----
[2025-05-28T15:22:34Z INFO  zabbix_api::client::client] getting auth session for user 'Admin'..
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] send post request to 'http://localhost:3080/api_jsonrpc.php'
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] ---[HTTP REQUEST]----
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] {"jsonrpc":"2.0","method":"user.login","params":{"password":"zabbix","username":"Admin"},"id":1}
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] ---[/HTTP REQUEST]----
[2025-05-28T15:22:34Z DEBUG reqwest::connect] starting new connection: http://localhost:3080/
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] ---[HTTP RESPONSE]----
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] {"jsonrpc":"2.0","result":"0930a6ef7dfe5d557e2ba863cabd2018","id":1}
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] ---[/HTTP RESPONSE]----
[2025-05-28T15:22:34Z INFO  zabbix_api::client::client] auth ok
[2025-05-28T15:22:34Z INFO  zabbix_api::client::client] creating host group 'oLJ0NxqbJTzoCAcgMIvYYo92QpSHyK'..
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] send post request to 'http://localhost:3080/api_jsonrpc.php'
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] ---[HTTP REQUEST]----
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] {"jsonrpc":"2.0","method":"hostgroup.create","params":{"name":"oLJ0NxqbJTzoCAcgMIvYYo92QpSHyK"},"id":1}
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] ---[/HTTP REQUEST]----
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] ---[HTTP RESPONSE]----
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] {"jsonrpc":"2.0","error":{"code":-32500,"message":"Application error.","data":"DBEXECUTE_ERROR"},"id":1}
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] ---[/HTTP RESPONSE]----
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::client] [response body]
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::client] {"jsonrpc":"2.0","error":{"code":-32500,"message":"Application error.","data":"DBEXECUTE_ERROR"},"id":1}
[2025-05-28T15:22:34Z DEBUG zabbix_api::client::client] [/response body]
[2025-05-28T15:22:34Z ERROR zabbix_api::client::client] ZabbixError { code: -32500, message: "Application error.", data: "DBEXECUTE_ERROR" }
Caused by: [zabbix error] code -32500, message 'Application error.', data: 'DBEXECUTE_ERROR' [/zabbix error]
[2025-05-28T15:22:34Z ERROR zabbix_api::tests::builder] host group create error: zabbix api call error

thread 'client::client::tests::create_host_with_cert' panicked at src/tests/builder.rs:88:17:
zabbix api call error
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    client::client::tests::create_host_with_cert

test result: FAILED. 20 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.50s

error: test failed, to rerun pass `--lib`
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.07s
   Doc-tests zabbix_api

running 19 tests
```

But it's ok after second attempt.

judu commented:
> ```
> [2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] ---[HTTP REQUEST]----
> [2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] {"jsonrpc":"2.0","method":"hostgroup.create","params":{"name":"oLJ0NxqbJTzoCAcgMIvYYo92QpSHyK"},"id":1}
> [2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] ---[/HTTP REQUEST]----
> [2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] ---[HTTP RESPONSE]----
> [2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] {"jsonrpc":"2.0","error":{"code":-32500,"message":"Application error.","data":"DBEXECUTE_ERROR"},"id":1}
> [2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] ---[/HTTP RESPONSE]----
> ```
> 
> But it's ok after second attempt.

I had the same error but on some other random request. As you said, the call succeeded when running the tests again. I have no idea where this error is coming from, but I attributed it to the dockered zabbix configuration. I actually did not have this error while running teste with v7, but started to get that error when I ran them against v6.

Let me try and get more logs from zabbix.

For my memory, here’s what I found on this error: https://support.zabbix.com/browse/ZBX-9916?page=com.atlassian.jira.plugin.system.issuetabpanels%3Achangehistory-tabpanel

lebe-dev commented on src/host/create.rs:
```diff
@@ -9,6 +10,38 @@ use crate::{
+    Cert { issuer: String, subject: String },
+}
+
+impl Serialize for TlsConfig {
```
@judu does it really need to implement Serialize? Automatic serialization is not enough?

Example: https://github.com/lebe-dev/zabbix-api-rs/blob/e025ff0cb6929d101615e6025eba73b9d787f13c/src/item/create.rs#L6

judu replied:
Well, with the choices I made, I had to implement it myself. Couldn’t find how to bend the derived Serializer to my needs :innocent: 

Anyway, after a second thought, I’ve amended my code so the structs sticks to the zabbix api. I just added some constructors to help and guide the users (and myself).

lebe-dev reviewed (changes requested):
Suggestions:
1. Rename argument to `tls_config`
2. Replace manual serialization with auto

lebe-dev commented:
> > ```
> > [2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] ---[HTTP REQUEST]----
> > [2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] {"jsonrpc":"2.0","method":"hostgroup.create","params":{"name":"oLJ0NxqbJTzoCAcgMIvYYo92QpSHyK"},"id":1}
> > [2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] ---[/HTTP REQUEST]----
> > [2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] ---[HTTP RESPONSE]----
> > [2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] {"jsonrpc":"2.0","error":{"code":-32500,"message":"Application error.","data":"DBEXECUTE_ERROR"},"id":1}
> > [2025-05-28T15:22:34Z DEBUG zabbix_api::client::post] ---[/HTTP RESPONSE]----
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > But it's ok after second attempt.
> 
> I had the same error but on some other random request. As you said, the call succeeded when running the tests again. I have no idea where this error is coming from, but I attributed it to the dockered zabbix configuration. I actually did not have this error while running teste with v7, but started to get that error when I ran them against v6.
> 
> Let me try and get more logs from zabbix.
> 
> For my memory, here’s what I found on this error: https://support.zabbix.com/browse/ZBX-9916?page=com.atlassian.jira.plugin.system.issuetabpanels%3Achangehistory-tabpanel

Thanks for the investigation, it seems not so sufficient for us :) But due to respect crate users we should mention ZBX-9916 somewhere in docs.

judu committed:
docs: check "host delete" operation

judu committed:
fix(host.create): templates field only needs template ids

Tried to create a host with a template list containing anything else
than "templateid", got this error:


```json
{
  "jsonrpc": "2.0",
  "error": {
    "code": -32602,
    "message": "Invalid params.",
    "data": "Invalid parameter \"/1/templates/2\": unexpected parameter \"host\"."
  },
  "id": 1
}
```

judu committed:
docs: document the weird zabbix behaviour during tests

judu commented:
Regarding the error in the tests, as promised, I investigated a bit more! I wanted to make extra sure the errors were *not* from my code. (I suspected my change of test/strings.rs to generate invalid names of sorts.)

I checked out the main branch and ran the integration tests:

```
running 17 tests
test client::client::tests::get_api_info ... ok
test client::client::tests::create_item ... FAILED
test client::client::tests::create_web_scenario ... FAILED
test client::client::tests::create_user_group ... ok
test client::client::tests::create_host_group_and_host ... ok
test client::client::tests::get_hosts_test ... ok
test client::client::tests::raw_api_call_test ... ok
test client::client::tests::create_trigger ... ok
test client::client::tests::get_host_groups_test ... ok
test trigger::model::trigger_tests::deserialize_test ... ok
test client::client::tests::session_should_be_returned ... ok
test client::client::tests::get_triggers_test ... ok
test client::client::tests::get_items_test ... ok
test client::client::user_tests::get_user_groups_test ... ok
test client::client::tests::get_webscenarios_test ... ok
test client::client::user_tests::get_users_test ... ok
test client::client::user_tests::create_user_test ... ok

failures:

---- client::client::tests::create_item stdout ----
[2025-05-30T21:28:34Z INFO  zabbix_api::client::client] getting auth session for user 'Admin'..
[2025-05-30T21:28:34Z DEBUG zabbix_api::client::post] send post request to 'http://localhost:3080/api_jsonrpc.php'
[2025-05-30T21:28:34Z DEBUG reqwest::connect] starting new connection: http://localhost:3080/
[2025-05-30T21:28:34Z DEBUG zabbix_api::client::post] ---[HTTP RESPONSE]----
[2025-05-30T21:28:34Z DEBUG zabbix_api::client::post] {"jsonrpc":"2.0","result":"c174a65c0cae5210ee283ef03c6859a6","id":1}
[2025-05-30T21:28:34Z DEBUG zabbix_api::client::post] ---[/HTTP RESPONSE]----
[2025-05-30T21:28:34Z INFO  zabbix_api::client::client] auth ok
[2025-05-30T21:28:34Z INFO  zabbix_api::client::client] creating host group 'TqUCJThasGl2AMNb'..
[2025-05-30T21:28:34Z DEBUG zabbix_api::client::post] send post request to 'http://localhost:3080/api_jsonrpc.php'
[2025-05-30T21:28:34Z DEBUG zabbix_api::client::post] ---[HTTP RESPONSE]----
[2025-05-30T21:28:34Z DEBUG zabbix_api::client::post] {"jsonrpc":"2.0","error":{"code":-32500,"message":"Application error.","data":"DBEXECUTE_ERROR"},"id":1}
[2025-05-30T21:28:34Z DEBUG zabbix_api::client::post] ---[/HTTP RESPONSE]----
[2025-05-30T21:28:34Z DEBUG zabbix_api::client::client] [response body]
[2025-05-30T21:28:34Z DEBUG zabbix_api::client::client] {"jsonrpc":"2.0","error":{"code":-32500,"message":"Application error.","data":"DBEXECUTE_ERROR"},"id":1}
[2025-05-30T21:28:34Z DEBUG zabbix_api::client::client] [/response body]
[2025-05-30T21:28:34Z ERROR zabbix_api::client::client] ZabbixError { code: -32500, message: "Application error.", data: "DBEXECUTE_ERROR" }
Caused by: [zabbix error] code -32500, message 'Application error.', data: 'DBEXECUTE_ERROR' [/zabbix error]
[2025-05-30T21:28:34Z ERROR zabbix_api::tests::builder] host group create error: zabbix api call error

thread 'client::client::tests::create_item' panicked at src/tests/builder.rs:86:17:
zabbix api call error
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- client::client::tests::create_web_scenario stdout ----
[2025-05-30T21:28:34Z INFO  zabbix_api::client::client] getting auth session for user 'Admin'..
[2025-05-30T21:28:34Z DEBUG zabbix_api::client::post] send post request to 'http://localhost:3080/api_jsonrpc.php'
[2025-05-30T21:28:34Z DEBUG reqwest::connect] starting new connection: http://localhost:3080/
[2025-05-30T21:28:34Z DEBUG zabbix_api::client::post] ---[HTTP RESPONSE]----
[2025-05-30T21:28:34Z DEBUG zabbix_api::client::post] {"jsonrpc":"2.0","result":"2e6482b7236bcf36e48a8acd1c194f34","id":1}
[2025-05-30T21:28:34Z DEBUG zabbix_api::client::post] ---[/HTTP RESPONSE]----
[2025-05-30T21:28:34Z INFO  zabbix_api::client::client] auth ok
[2025-05-30T21:28:34Z INFO  zabbix_api::client::client] creating host group '9bBtvRrK'..
[2025-05-30T21:28:34Z DEBUG zabbix_api::client::post] send post request to 'http://localhost:3080/api_jsonrpc.php'
[2025-05-30T21:28:34Z DEBUG zabbix_api::client::post] ---[HTTP RESPONSE]----
[2025-05-30T21:28:34Z DEBUG zabbix_api::client::post] {"jsonrpc":"2.0","error":{"code":-32500,"message":"Application error.","data":"DBEXECUTE_ERROR"},"id":1}
[2025-05-30T21:28:34Z DEBUG zabbix_api::client::post] ---[/HTTP RESPONSE]----
[2025-05-30T21:28:34Z DEBUG zabbix_api::client::client] [response body]
[2025-05-30T21:28:34Z DEBUG zabbix_api::client::client] {"jsonrpc":"2.0","error":{"code":-32500,"message":"Application error.","data":"DBEXECUTE_ERROR"},"id":1}
[2025-05-30T21:28:34Z DEBUG zabbix_api::client::client] [/response body]
[2025-05-30T21:28:34Z ERROR zabbix_api::client::client] ZabbixError { code: -32500, message: "Application error.", data: "DBEXECUTE_ERROR" }
Caused by: [zabbix error] code -32500, message 'Application error.', data: 'DBEXECUTE_ERROR' [/zabbix error]
[2025-05-30T21:28:34Z ERROR zabbix_api::tests::builder] host group create error: zabbix api call error

thread 'client::client::tests::create_web_scenario' panicked at src/tests/builder.rs:86:17:
zabbix api call error


failures:
    client::client::tests::create_item
    client::client::tests::create_web_scenario

test result: FAILED. 15 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.81s
```

It seems to be the same error as yours.

The container got me these logs that matched the failed tests:

```
zabbix-web-nginx-mysql-1  | [30-May-2025 21:28:34] WARNING: [pool zabbix] child 55 said into stderr: "[31-May-2025 00:28:34 Europe/Riga] PHP Warning:  Error in query [INSERT INTO ids (table_name,field_name,nextid) VALUES ('hstgrp','groupid',22)] [Deadlock found when trying to get lo
ck; try restarting transaction] in /usr/share/zabbix/include/db.inc.php on line 249"
zabbix-web-nginx-mysql-1  | [30-May-2025 21:28:34] WARNING: [pool zabbix] child 53 said into stderr: "[31-May-2025 00:28:34 Europe/Riga] PHP Warning:  Error in query [INSERT INTO ids (table_name,field_name,nextid) VALUES ('hstgrp','groupid',22)] [Deadlock found when trying to get lo
ck; try restarting transaction] in /usr/share/zabbix/include/db.inc.php on line 249"
```

It looks like a race condition, maybe the tests running too fast?

As you mentionned, running them again on the same zabbix threw no error. I tried it multiple times and it looks like it only happens on a fresh container run. It happens on `main` and on my branch.

I’m adding something to the doc, please tell me if it’s fine for you!

judu commented:
Oh, beware! I added [this commit][https://github.com/lebe-dev/zabbix-api-rs/pull/17/commits/7e2f75fc1e866358cc78c11d39e73637900e05c9) that changes the CreateHostRequest to use `Vec<TemplateId>` instead of `Vec<Template>`. Since putting anything more than a "templateid" in the object got me an error, I think that it’s safe to change as it seems no one was using it.
